src/librustc_mir/transform/qualify_min_const_fn.rs

```diff
@@ -57,217 +58,226 @@ pub fn is_min_const_fn(tcx: TyCtxt<'tcx>, def_id: DefId, body: &'a Body<'tcx>) -
         }
     }
 
-    for local in &body.local_decls {
-        check_ty(tcx, local.ty, local.source_info.span, def_id)?;
-    }
-    // impl trait is gone in MIR, so check the return type manually
-    check_ty(
+    let mut visitor = IsMinConstFn {
         tcx,
-        tcx.fn_sig(def_id).output().skip_binder(),
-        body.local_decls.iter().next().unwrap().source_info.span,
+        body,
         def_id,
-    )?;
+        span: body.span,
+    };
 
-    for bb in body.basic_blocks() {
-        check_terminator(tcx, body, def_id, bb.terminator())?;
-        for stmt in &bb.statements {
-            check_statement(tcx, body, def_id, stmt)?;
+    // We could do most of this simply by calling `visit_body`. However, we change the order since
+    // we want to check each parameter **before** the function body, and the terminator **before**
+    // the statements.
+    for (local, decl) in body.local_decls.iter_enumerated() {
+        visitor.visit_local_decl(local, decl)?;
+    }
+
+    // impl trait is gone in MIR, so check the return type manually
+    visitor.span = body.local_decls[RETURN_PLACE].source_info.span;
+    let ret_ty_ctx = TyContext::ReturnTy(body.local_decls.iter().next().unwrap().source_info);
+    visitor.visit_ty(tcx.fn_sig(def_id).output().skip_binder(), ret_ty_ctx)?;
+
+    for (bb, block) in body.basic_blocks().iter_enumerated() {
+        let mut loc = body.terminator_loc(bb);
+        visitor.visit_terminator(block.terminator(), loc)?;
+        for (idx, stmt) in block.statements.iter().enumerate() {
+            loc.statement_index = idx;
+            visitor.visit_statement(stmt, loc)?;
         }
     }
+
     Ok(())
 }
 
-fn check_ty(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, span: Span, fn_def_id: DefId) -> McfResult {
-    for ty in ty.walk() {
-        match ty.kind {
-            ty::Ref(_, _, hir::Mutability::MutMutable) => return Err((
-                span,
-                "mutable references in const fn are unstable".into(),
-            )),
-            ty::Opaque(..) => return Err((span, "`impl Trait` in const fn is unstable".into())),
-            ty::FnPtr(..) => {
-                if !tcx.const_fn_is_allowed_fn_ptr(fn_def_id) {
-                    return Err((span, "function pointers in const fn are unstable".into()))
+struct IsMinConstFn<'mir, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    span: Span,
+    body: &'mir Body<'tcx>,
+}
+
+impl Visitor<'tcx, McfResult> for IsMinConstFn<'mir, 'tcx> {
```

Comment:
I actually liked the fact that we didn't use a visitor for this code cause visitors are pretty fragile when it comes to adding new things to a construct. In particular, if the visitor gains new provided methods then chances are they won't be modified here and so we may accidentally allow new things on stable... how do we prevent this? cc @oli-obk

Reply:
I think the problem that you're describing is mostly solved with exhaustive matching on `Rvalue`, `StatementKind` etc. This type of defensive coding is a requirement for both when using `Visitor` and when manually recursing. Can you give an example where this kind of breakage would occur in one but not the other?

The downside of manual recursion is that it's easier to forget to call `visit_operand` or `visit_place` on something. The `super` methods put this code in one place, making it easier to audit for correctness. You don't have to clutter your code by matching all the fields of a `TerminatorKind::Call` and manually calling your equivalent of `visit`.

Reply:
> I think the problem that you're describing is mostly solved with exhaustive matching on `Rvalue`, `StatementKind` etc.

Sure, that works well when you are already matching on a specific enum and another variant is added, but what if a new enum or struct is added altogether?

> Can you give an example where this kind of breakage would occur in one but not the other?

I suspect it is more likely with manual recursion that we would notice that a new field has been added for example...

> The downside of manual recursion is that it's easier to forget to call `visit_operand` or `visit_place` on something.

...but this is a fair point; maybe I'm overly paranoid. :)

Reply:
I think a better critique of `Visitor` APIs is that they encourage keeping a lot of ancillary state in the implementer instead of explicitly passing data down via function parameters. This is exacerbated by by the fact that `Visitor` takes `&mut self`. Also, because the order of traversal is defined in the `super` methods, you need to check elsewhere to ensure that things are called in the order you expect. This is often a problem when updating a `span`.

Reply:
Here's a suggestion: Let's modify the visitor documentation to suggest that the developer first add the new visitor method *without a provided method*; this will make errors all over the place and the developer will need to consider them all. Once they have been considered, the method can become provided.

Reply:
@ecstatic-morse I agree with that critique *as well*. :) I think [SYB](https://www.microsoft.com/en-us/research/wp-content/uploads/2003/01/hmap.pdf) in the style of [lens](https://hackage.haskell.org/package/lens-4.15.1/docs/Control-Lens-Plated.html) is neater for that reason, but we don't have generic closures unfortunately.